### PR TITLE
REF: split out code writing logic

### PR DESCRIPTION
**Pull request by mzeitlin11**
The goal of this PR is to start separating out the actual generation of rust code from the rest of the code generation logic.  All logic for writing rust code should now live in `/templates`. The plan is for the beast that is `codegen.rs` to not have to worry about the actual printing of the code, making it a bit simpler.

While this pr also uses more `format!` style code to write the rust code, the purpose of this PR is to gain the benefits of starting to decouple this logic rather than decide on a specific code writing mechanism. There may be better solutions for this in the future, hopefully isolating all this code in `templates/` will make any future rust code generation changes simpler and less invasive.

There are also some other tangential benefits:
- Code writing is consolidated: most of handling struct and enum writing now is handled in a single place which should make future extensions easier
- Dependencies of items are becoming clearer, eg exactly what information is needed for a struct or enum to be written.

Note that right now most (or all :)) types are still `String` and their friends. This PR makes an in-progress branch much easier that is designed to start more strongly typing things (differentiate between strings which now represent various different concepts from a `RustType`, what identifier names are, raw strings specified in the stripe spec, and more). 

Since hopefully there is a lot of future code generation simplification coming, much of this code will likely end up changing, so there are some cleanups I didn't want to spend time worrying about yet  (which is one reason why there is some wanton cloning since I didn't want to think about lifetimes :) )

#### Generated code changes
The diff in the generated code is easier to look at with the ignore whitespace setting. The changes in the generated code should all be whitespace or doc changes caused by the following general changes:

- The comment for the `deleted` field name was changed since we now format it like other doc comments
- The comment for all structs now ends with a period since all are formatted in the same place
- `impl Paginate` unfortunately had to be moved to a different place in the printed file order when some logic was consolidated
- We use less code worrying about formatting since we'll run `rustfmt` later anyway. Beyond code simplification, this has the added benefit of (hopefully) reducing formatting related diffs in the future since we can gradually defer more and more to `rustfmt` and have fewer changes related to adding newlines and tabs. Unfortunately, this is annoying as long as there are unused generated files because they don't get formatted. However, this makes it clearer at a glance what files are unused :)
- Because of https://github.com/rust-lang/rustfmt/issues/3863 and some of the related issues mentioned in that chain, some rare extremely long lines /variable names have tab-related spacing changes since they aren't being formatted

### CODE REVIEW

File: src/resources/generated/tax_rate.rs [hunk 1]
```diff
@@ -159,14 +159,14 @@ pub struct CreateTaxRate<'a> {
 }
 
 impl<'a> CreateTaxRate<'a> {
-    pub fn new(display_name: &'a str, percentage: f64) -> Self {
+    pub fn new(display_name: &'a str, inclusive: bool, percentage: f64) -> Self {
```
Comment (mzeitlin11) [inline src/resources/generated/tax_rate.rs:162]: This is a breaking change in the generated code I need to fix before removing this pr from draft

**Comment by arlyon** [comment]
Hey! Took some time off to recharge. Gonna check this out today. Hope you had a nice break!

**Comment by mzeitlin11** [comment]
> Hey! Took some time off to recharge. Gonna check this out today. Hope you had a nice break!

I did thanks - glad you could too! Also am working on another branch built on this which might be a bit more interesting :) . Hopefully will have a draft up today / tomorrow so no rush on reviewing this

**Comment by arlyon** [comment]
Hi! You can rebase this branch now, the code coverage PR is merged :)

**Comment by arlyon** [comment]
Lets get this in and we can start working seriously on #314. I am open to merging that PR into a new branch (perhaps `next` which we can use for prerelease versions. Approving with that in mind, and I will merge once CI is good.

**Review by arlyon — APPROVED** [review]
You can address this comment if you'd like. Not a huge deal.

**Comment by mzeitlin11** [comment]
Sorry for the delay here - going to put this into draft for now until a bit more progress made on #314. I think some of the changes there will influence this pr.

**Comment by mzeitlin11** [comment]
Closing as this does not appear to be a useful precursor

**Pull request closed without merging**